server/embed/etcd.go

```diff
@@ -402,6 +403,16 @@ func (e *Etcd) Close() {
 		lg.Sync()
 	}()
 
+	// in case the server did not start or did not stop on its own stop it directly
+	if e.Server != nil {
+		select {
+		case <-e.Server.ReadyNotify():
+		case <-e.Server.StopNotify():
+		default:
+			e.Server.Stop()
+		}
+	}
```

Comment:
Please see https://github.com/etcd-io/etcd/pull/16758#discussion_r1375469179.

Suggest to review & evaluate this change (I mean the changes in `embed/etcd.go` file) separately. There might be some subtle cases we need to be more careful. We can discuss & evaluate it separately.

Reply:
I believe the extra changes were required for the [partial cluster test](https://github.com/etcd-io/etcd/pull/16754#issuecomment-1761332546) that you wanted me too add. I can try again but all changes that I did were required to make the tests pass.

Reply:
> I believe the extra changes were required for the [partial cluster test](https://github.com/etcd-io/etcd/pull/16754#issuecomment-1761332546) that you wanted me too add

Could you move the change in` embed/etcd.go` file and the related case(s) into a separate PR. I will take a closer look separately.

Reply:
If I only take the commits
- 605ed7c5cc19fc97fbf14a8f5018e5e4bfb01740
- 537b1aafd118410a9c2ffb18ec380fc5f0795ca8
- 5660097b141ed6109b6a8adeaae5c5b8ec507b29

the test (`GOOS=linux GOARCH=amd64 CPU=2 make test-integration`) fails due to a data race condition:
```
==================
WARNING: DATA RACE
Write at 0x00c002ae3090 by goroutine 1117:
  runtime.closechan()
      /var/lib/snapd/snap/go_121/10389/src/runtime/chan.go:357 +0x0
  go.etcd.io/etcd/server/v3/embed.(*Etcd).Close()
      /home/hendrik/projects/github/Hendrik-H/etcd/server/embed/etcd.go:454 +0xcee
  go.etcd.io/etcd/tests/v3/integration/embed_test.testEmbedEtcdClusterGracefulStop.func1()
      /home/hendrik/projects/github/Hendrik-H/etcd/tests/integration/embed/embed_test.go:254 +0xd7

Previous read at 0x00c002ae3090 by goroutine 1113:
  runtime.chansend()
      /var/lib/snapd/snap/go_121/10389/src/runtime/chan.go:160 +0x0
  go.etcd.io/etcd/server/v3/embed.(*Etcd).errHandler()
      /home/hendrik/projects/github/Hendrik-H/etcd/server/embed/etcd.go:865 +0x2aa
  go.etcd.io/etcd/server/v3/embed.(*Etcd).serveClients.func1()
      /home/hendrik/projects/github/Hendrik-H/etcd/server/embed/etcd.go:769 +0x19c
  go.etcd.io/etcd/server/v3/embed.(*Etcd).serveClients.func3()
      /home/hendrik/projects/github/Hendrik-H/etcd/server/embed/etcd.go:770 +0x41

Goroutine 1117 (running) created at:
  go.etcd.io/etcd/tests/v3/integration/embed_test.testEmbedEtcdClusterGracefulStop()
      /home/hendrik/projects/github/Hendrik-H/etcd/tests/integration/embed/embed_test.go:251 +0xa44
  go.etcd.io/etcd/tests/v3/integration/embed_test.TestEmbedEtcdClusterGracefulStopInsecure()
      /home/hendrik/projects/github/Hendrik-H/etcd/tests/integration/embed/embed_test.go:198 +0x28
  testing.tRunner()
      /var/lib/snapd/snap/go_121/10389/src/testing/testing.go:1595 +0x238
  testing.(*T).Run.func1()
      /var/lib/snapd/snap/go_121/10389/src/testing/testing.go:1648 +0x44

Goroutine 1113 (finished) created at:
  go.etcd.io/etcd/server/v3/embed.(*Etcd).serveClients()
      /home/hendrik/projects/github/Hendrik-H/etcd/server/embed/etcd.go:768 +0xa71
  go.etcd.io/etcd/server/v3/embed.StartEtcd()
      /home/hendrik/projects/github/Hendrik-H/etcd/server/embed/etcd.go:274 +0x2597
  go.etcd.io/etcd/tests/v3/integration/embed_test.testEmbedEtcdClusterGracefulStop()
      /home/hendrik/projects/github/Hendrik-H/etcd/tests/integration/embed/embed_test.go:240 +0x6c5
  go.etcd.io/etcd/tests/v3/integration/embed_test.TestEmbedEtcdClusterGracefulStopInsecure()
      /home/hendrik/projects/github/Hendrik-H/etcd/tests/integration/embed/embed_test.go:198 +0x28
  testing.tRunner()
      /var/lib/snapd/snap/go_121/10389/src/testing/testing.go:1595 +0x238
  testing.(*T).Run.func1()
      /var/lib/snapd/snap/go_121/10389/src/testing/testing.go:1648 +0x44
==================
```
This is fixed by adding commit 6bb13495ab393feb2548f756dc2c59654d5e030f.

So if we only want to have 537b1aafd118410a9c2ffb18ec380fc5f0795ca8 then it looks like I can unfortunately not provide a test case.

Reply:
Hi @ahrtr , how can we get further with this fix? As stated above if I don't take all changes the test case do not pass.